### DISCUSSION TITLE
Consider trading usage of Reflection.Emit with Source Generator in System.Text.Json

### DISCUSSION THREAD

**Discussion by am11** [dotnet/runtime#40903] [General] [answered]
After witnessing a bug with runtime IL emission in System.Text.Json: https://github.com/dotnet/runtime/issues/40878, I read about C#'s new feature, [Source Generator](https://devblogs.microsoft.com/dotnet/introducing-c-source-generators/), and was wondering if there are considerations to completely switch to Source Generator in S.T.J? In addition to the fact that it will make things AOT-friendly (and performant), it would also help catching the typecasting issues during the compile time.

An alternative idea is https://github.com/dotnet/runtime/issues/40620 - to verify generated IL on runtime.

However, I think S.T.J is the top candidate to get SourceGenerator-based implementation, since its description states:

> The System.Text.Json namespace provides <ins>high-performance, low-allocating</ins>, and standards-compliant capabilities to process JavaScript Object Notation (JSON), which includes serializing objects to JSON text and deserializing JSON text to objects, with UTF-8 support built-in.

**Comment by gfoidl** [comment, marked as answer]
--> https://github.com/dotnet/runtimelab/pull/16 (there's an experminent over there).

**Reply by am11** [reply]
Noice 😎
this is exactly what I was hoping to find.
